### ISSUE TITLE
NullReferenceException when trying to GET /Item/Images/{filename}

### ISSUE THREAD

**Issue by caribgit** [NCRCounterpointAPI/APIGuide#97]
Have checked that an image filename exists in the <COMPANY>\Configuration\ItemImages folder

When calling the image using the GET /Item/Images/{filename} call, we get this:

Response Status
Error CodeNullReferenceExceptionMessageObject reference not set to an instance of an object.Stack Trace[ItemImageRequest: 2/8/2018 10:54:52 PM]: [REQUEST: {filename:2ND BC-65DBST.JPG}] System.NullReferenceException: Object reference not set to an instance of an object. at CPAPI.Services.CPAPICompanyService.get_CompanyName() at CPAPI.Services.ItemImageService.Get(ItemImageRequest request) at ServiceStack.Host.ServiceRunner`1.Execute(IRequest request, Object instance, TRequest requestDto)Errors

**Comment by james-prickett** [comment]
I just want to add that even thought the ITEM_NO from the example above has a space in it, we found that this error happens regardless to us even if the ITEM_NO has a space in it.

**Comment by jbbehar** [comment]
Per https://github.com/NCRCounterpointAPI/APIGuide/issues/29 have you tried using .jpg (lower case file extension)?

**Comment by james-prickett** [comment]
We're using the filename returned by the GET /Item/{ItemNo}/Images service.   In this example, the filename returned has a lower case extension.   The file exists on the NCR server too... with lower case extension.

**Comment by MikeRitchie** [comment]
This piece of your error makes me wonder if there's a config issue. Most API calls just hit the DB, but a few, like images, hit the TLD as well. Can you check the setup for the company in the API management console (logged in as admin)? Also, can you verify a call like GET /Item/{ItemNo} works as well? It sounds like GET/Item/{ItemNo}/Images works on this  same system without issue? If so that indicates the TLD config should be good.

Also, could you directly email me the logfile containing this request? I did a quick postman test and was able to successfully retrieve an image.

**Comment by caribgit** [comment]
Thank you Mike. It seems to be related to the hostname being used. We ran this: https://localhost:52000/Item/Images/2ND%20T-49F.jpg and it fails, but if we run https://asd-svr-02:52000/Item/Images/2ND%20T-49F.jpg it works. 

Because the client does not have a static IP, we are connecting to the API remotely using a dynamic DNS hostname and when we use this hostname it does not work, similar to what happens if we use the locahost as the hostname. 

Can the dynamic DNS hostname be also added to the configuration and if so where would this be done?

**Comment by james-prickett** [comment]
Regarding  GET /Item/{ItemNo} and GET /Item/{ItemNo}/Images,  I'm calling the web services from a PHP client and both are working for me.  Just the call to get the actual image.   Looks like DNS may be the issue but wanted to follow up anyway..

```php
php > $item = $ncrService->getItem('2ND T-49F');


php >
php >
php > var_dump($item);
object(stdClass)#19 (79) {
  ["ITEM_NO"]=>
  string(9) "2ND T-49F"
  ["DESCR"]=>
  string(26) "REACH-IN FREEZER 2 SS-DOOR"
  ["DESCR_UPR"]=>
  string(26) "REACH-IN FREEZER 2 SS-DOOR"
  ["LONG_DESCR"]=>
  string(26) "REACH-IN FREEZER 2 SS DOOR"
  ["LONG_DESCR_UPR"]=>
  string(26) "REACH-IN FREEZER 2 SS DOOR"
  ["ADDL_DESCR_1"]=>
  string(18) "SERIAL NO. 8894771"
  ["ITEM_TYP"]=>
  string(1) "I"
  ["CATEG_COD"]=>
  string(8) "2ND HAND"
  ["SUBCAT_COD"]=>
  string(3) "REF"
  ["ACCT_COD"]=>
  string(3) "100"
  ["TRK_METH"]=>
  string(1) "N"
  ["STAT"]=>
  string(1) "A"
  ["IS_TXBL"]=>

```

**Comment by MikeRitchie** [comment]
I'm a bit puzzled that localhost fails but using the machine name works. I assume with localhost you are working off of the machine the server is on? There's nothing related to DNS that I believe would fall on the API side. The API server simply listens on the configured port for incoming connections. DNS resolution, etc. all happens prior to that connection being made.

My only other guess would be that permissions could be at play. Perhaps the account the API service is running under has permission to list files (so the images call works), but can't access the actual file. This seems unlikely based off of the DNS data you provided but may be worth checking.

**Comment by caribgit** [comment]
Yes both attempts, one using localhost and the other using asd-srv-02 are done on the local machine. When you made your postman request, what hostname did you use and what user did you use? You said you successfully retrieved an image from their system, correct?

As regards the permissions you mention,  the API is configured to run under "Local System Account".

**Comment by MikeRitchie** [comment]
Correct, I can retrieve an image with localhost or by machine name. Is your TLD on a remote machine? It could be that the "Local System Account" can't pull the image off of the remote TLD machine if the remote TLD doesn't allow that account to read the files. The simplest check for this may be to open up the permissions fully on the TLD share and see if that resolves the issue. If it does then you could lock back down to just the necessary permissions.

I wonder if it could also be an SSL issue. Sometimes I think the machine or client may block on localhost where it doesn't with machine name. Setting up the self-signed certificate as a trusted root authority could help with this as well.

**Comment by MikeRitchie** [comment]
You might also try hitting https://localhost:52000 in chrome, and if chrome shows a warning, hit advanced>proceed to tell Chrome to trust the website. This will probably only help if you're using the Postman Chrome web app, however.

**Comment by caribgit** [comment]
I don't have an issue accessing https://localhost:52000/, no warnings. But this is because when I type that  URL in it automatically redirects to https://asd-svr-02:52000 whereas if I type a full URL such as https://localhost:52000/Item/Images/2ND%20T-49F.jpg to get the image it does not redirect and I see a warning that the connection is not secure and then it shows the Response status with  NullReferenceException etc.

**Comment by mbantug** [comment]
The certificate we install by default for TLS is tied to machine name and not localhost.  That's probably why it works in some cases where it automatically redirects you.

If you need localhost to work, you could generate your own cert and include both machinename and localhost in the SAN list.
